### ISSUE TITLE
[Linux][arm] CoreFx test System.Transactions.Local.Tests hangs and more

### ISSUE THREAD

**Issue by echesakovMSFT** [dotnet/coreclr#17858]
CoreFx test `System.Transactions.Local.Tests` hangs during `System.Transactions.Tests.AsyncTransactionScopeTests.VerifyBYOTAsyncTSNestedAsync` on Ubuntu/arm with both Checked and Release coreclr
(as an example, https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/arm_cross_checked_ubuntu_corefx_baseline_tst_prtest/1/consoleText)

```
xUnit.net console test runner (32-bit .NET Core)
Copyright (C) 2014 Outercurve Foundation.

Discovering: System.Transactions.Local.Tests
Discovered:  System.Transactions.Local.Tests
Starting:    System.Transactions.Local.Tests
   System.Transactions.Tests.AsyncTransactionScopeTests.VerifyBYOTAsyncTSNestedAsync [STARTING]
```
What is interesting - if you keep at least one core busy (e.g. by running `stress --cpu 1 --timeout 600`) the test always passes (I have a quad-core ARMv7 cpu) and you get

```
xUnit.net console test runner (32-bit .NET Core)
Copyright (C) 2014 Outercurve Foundation.

Discovering: System.Transactions.Local.Tests
Discovered:  System.Transactions.Local.Tests
Starting:    System.Transactions.Local.Tests
   System.Transactions.Tests.AsyncTransactionScopeTests.VerifyBYOTAsyncTSNestedAsync [STARTING]
   System.Transactions.Tests.AsyncTransactionScopeTests.VerifyBYOTAsyncTSNestedAsync [FINISHED] Time: 0.2283526s
Finished:    System.Transactions.Local.Tests

=== TEST EXECUTION SUMMARY ===
   System.Transactions.Local.Tests  Total: 1, Errors: 0, Failed: 0, Skipped: 0, Time: 0.935s
```

Steps to reproduce is:
1. Build (Checked or Release) coreclr
2. Build Release corefx tests
3. Run something like this inside System.Transactions.Local.Tests folder:
```
coreclr/_/fx-Release/bin/testhost/netcoreapp-Linux-Release-arm/dotnet xunit.console.netcore.exe System.Transactions.Local.Tests.dll  -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapptests -notrait category=nonlinuxtests  -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing -method System.Transactions.Tests.AsyncTransactionScopeTests.VerifyBYOTAsyncTSNestedAsync -showprogress
```
It doesn't seem to be reproducible on Windows/arm though, so could be a Linux specific issue.

**More:** I also tend to believe that this is **not** a test specific. The problem just manifests itself as a deadlock in this particular scenario, but in other scenarios it could be just an unnecessary waiting. For example, running `System.Collections.Tests` test suite on average (in 10 test runs) takes 62.5s on my machine (with Release coreclr). However, if I do the same trick by running `stress` in a separate ssh window the tests will take on average 38s. This is also true for `System.Runtime.Tests` - 92s vs 55s.
We also confirmed with @BruceForstall that this happens with Checked coreclr on a different machine (same kernel and cpu though).

@janvorli  Is this something you would be interested to look into and see if it can be reproduced on your hardware? 
/cc @RussKeldorph

**Comment by echesakovMSFT** [comment]
**Update**: As suggested by @RussKeldorph I have tried to disable CPU scaling and force all cores to run at max performance. That seems explaining difference in running times with and without `stress` in background.

**Comment by BruceForstall** [comment]
@stephentoub @danmosemsft Is there a corefx dev who can investigate this? It seems like possibly a threading/locking issue in the test (or product).

**Comment by danmosemsft** [comment]
What is the evidence it is a deadlock rather than just slow? I do not see that above.

**Comment by BruceForstall** [comment]
We see the System.Transactions.Local.Tests test "hang" for a long period of time (meaning, it never finishes). When it passes, it takes less than a second.

**Comment by danmosemsft** [comment]
Gotcha. Per https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md, owners are @jimcarley, @qizhanMS, @dmetzgar.

Could you help them get a dumpfile/symbols to start?

**Comment by echesakovMSFT** [comment]
I have attached gdb backtraces and decoded top managed code frames: [17858.txt](https://github.com/dotnet/coreclr/files/1975958/17858.txt)

**Comment by danmosemsft** [comment]
Is it possible to get full managed stacks, eg `clrstack -all -a -f` in SOS?

**Comment by qizhanMS** [comment]
In the gdb trace, there is no System.Transaction function in the call stacks. It looks like either test didn't get started, or already finished.

**Comment by BruceForstall** [comment]
> Is it possible to get full managed stacks, eg clrstack -all -a -f in SOS?

Unfortunately, no. We can't run SOS on these Ubuntu/arm boxes because it requires LLDB, and we haven't been able to get LLDB to run.

The GDBJIT tool (built into CLR) might help, and we should try to enable it.

> In the gdb trace, there is no System.Transaction function in the call stacks. It looks like either test didn't get started, or already finished.

That may be because we don't have managed code on the stacks. @echesakovMSFT did print out the top frame. Possibly System.Transactions is below that top frame.

**Comment by RussKeldorph** [comment]
@qizhanMS Are you looking into this?  Any update?

@echesakov Now that you have lldb working, can you try `clrstack -all -a -f` per @danmosemsft?

**Comment by echesakovMSFT** [comment]
@RussKeldorph  Sure
@qizhanMS @danmosemsft Please find attached [clrstack -all -a -f.txt](https://github.com/dotnet/coreclr/files/1989762/clrstack-all-a-f.txt)

**Comment by RussKeldorph** [comment]
@qizhanMS @danmosemsft Any ideas based on `clrstack` output?  We can set you up with the repro machine if you could investigate.

**Comment by stephentoub** [comment]
@RussKeldorph, do you see hangs still if you add `-parallel none` to the xunit command line? Or if invoking with msbuild, `/p:XunitOptions="-parallel none"`?  My hypothesis is that this is a test issue, that xunit's SynchronizationContext is limiting the number of threads used to run tests, and these tests are doing blocking calls (e.g. Task.Result/.Wait/.WaitAll/etc) waiting for async operations to complete, but those async operations are scheduling back to that same context and causing deadlocks.

**Comment by echesakovMSFT** [comment]
@stephentoub It did hang even before with `-parallel none`

**Comment by danmosemsft** [comment]
Great that lldb is working! 

To eliminate @stephentoub 's suggestion, it might be worth adding `.ConfigureAwait(false)` onto all awaits - I 16 hits to the regex `await (?!(.*ConfigureAwait))` from https://github.com/dotnet/corefx/issues/29483. Also to try running just this one single test, either by adding `/p:xunitoptions="-method System.Transactions.Tests.AsyncTransactionScopeTests.VerifyBYOTAsyncTSNestedAsync` or by commenting out the rest - perhaps the callstacks will be clearer.

**Comment by stephentoub** [comment]
> It did hang even before with -parallel none

Hmm, ok, thanks.  Could you try with #29665?  If `-parallel none` didn't fix it, #29665 shouldn't either, but I just want to make sure before folks go chasing other things.

**Comment by echesakovMSFT** [comment]
Once dotnet/corefx#29665 got merged I will enable the test assembly in https://github.com/dotnet/coreclr/blob/d31fdcaa70cc6b9a969a4e2158eb545aee9c624b/tests/arm/corefx_linux_test_exclusions.txt#L15
